### PR TITLE
feat(dataset): move data iter creation to dataset class

### LINKED ISSUE #188: Regression: custom batches no longer supported
**Describe the bug**
In `joeynmt<2.0.0`, the `TrainManager` used to receive a `batch_class` to handle custom batches.
In `joeynmt=2.0.0`, it is no longer needed, as there is the `make_data_iter` which calls a `collate_fn` responsible to return whatever.

This created a regression - it used to be possible to use this repository as a library - but now one has to make hard-coded changes in the code in order to have a custom `make_data_iter`.

It is hard coded in [training](https://github.com/joeynmt/joeynmt/blob/main/joeynmt/training.py#L390) and [prediction](https://github.com/joeynmt/joeynmt/blob/main/joeynmt/prediction.py#L112)

**Possible Solutions:**
For training:
1. to have `data_iter` callable passed to the `TrainManager`
2. in the train manager, have a method so we can extend it:
```py
def make_data_iter(self, **kwargs):
  return make_data_iter(**kwargs)
```

For prediction:
1. pass this as an argument to the `predict` method?

----

What solutions would you prefer/approve if any?

### PR DESCRIPTION
**Pull request by AmitMY**
fixes #188 

- https://github.com/joeynmt/joeynmt/pull/189/commits/7a03d297373bdcb8a154e42955ab08ed0660e119 Minimally moved the `make_data_iter` to the dataset class, such that it can be overridden.  
- https://github.com/joeynmt/joeynmt/pull/189/commits/0fb08b8d5e276d026528c2c36c49217618119854 Moved `load_data` to a different file (`data_loader.py`) to prevent a circular dependency between `data` and `datasets`. 

(This was not tested for my exact use case)

### CODE REVIEW

File: joeynmt/datasets.py [hunk 1]
```diff
@@ -106,6 +107,9 @@ def trg(self) -> List[str]:
         """get detokenized preprocessed data in trg language."""
         return self.get_list(self.trg_lang, tokenized=False) if self.has_trg else []
 
+    def make_iter(self, **kwargs):
+        return make_data_iter(dataset=self, **kwargs)
```
Comment (AmitMY) [inline joeynmt/datasets.py:111]: My custom solution for my continuous case would then be:

1. Define a custom batch class
```python
class SignBatch(Batch):
    def __init__(self, src_length, **kwargs):
        super().__init__(src_length=src_length, **kwargs)
        self.src_mask = self._pose_mask(src_length)

    def _pose_mask(self, pose_length: torch.Tensor) -> torch.Tensor:
        max_len = pose_length.max().item()
        return torch.arange(max_len)[None, :] < pose_length[:, None]
```

2. Define a custom collation function  to `zero_pad` and instantiate the batch:

```python
def collate_fn(
        batch: List[Dict[str, Any]],
        pad_index: int = PAD_ID,
        device: torch.device = CPU_DEVICE,
        has_trg: bool = True,
        is_train: bool = True) -> Batch:
    collated = zero_pad_collator(batch)

    return SignBatch(
        src=collated["src"],
        src_length=collated["src_length"],
        trg=collated["trg"],
        trg_length=collated["trg_length"],
        device=device,
        pad_index=pad_index,
        has_trg=has_trg,
        is_train=is_train,
    )
```
3. Overriding the dataset `make_iter` method:
```python
def make_iter(self, pad_index: int = PAD_ID,
                  device: torch.device = CPU_DEVICE,
                  **kwargs) -> DataLoader:
    data_loader = super().make_iter(**kwargs)
    dataset = data_loader.dataset

    # data iterator
    return DataLoader(
        dataset=dataset,
        batch_sampler=data_loader.batch_sampler,
        collate_fn=partial(
            collate_fn,
            pad_index=pad_index,
            device=device,
            has_trg=True,
            is_train=dataset.split == "train",
        ),
        num_workers=data_loader.num_workers,
    )
```

Comment (juliakreutzer) [reply joeynmt/datasets.py:111]: thanks for this elaboration @AmitMY, this is very helpful.

**Review by juliakreutzer — CHANGES_REQUESTED** [review]
Looks very good, thank you @AmitMY. I agree with @may- 's comment though: three classes for data-related functions (dataset.py, data.py and data_loader.py) might be confusing. What do you think about her suggestion?

**Comment by may-** [comment]
@AmitMY Thank you for your contribution!

@juliakreutzer, I've confirmed the changes.

**Review by juliakreutzer — APPROVED** [review]
Awesome, thank you so much!

**Pull request merged**